### PR TITLE
Auto detect first variant for CPU and MEM

### LINKED ISSUE #21: Find machine parameters automatically
A lot of the parameters that you can supply could be found by the script automatically. Like:

- CPU Chips [integer [1,)]
- CPU Threads [integer [1,)]
- CPU Cores [integer [1,)]
- CPU Frequency [integer [1,)]
- Architecture [str]
- CPU Make [str]
- Release year [int]
- RAM [integer (0,]]
- TDP [integer (0,]]

We could probably just populate them when running the script. Or have a parameter `--auto`. 
ChatGPT gives me a good starting point:
```
import os
import subprocess
import re

def get_cpu_info():
    # CPU Chips, Threads, Cores, and Frequency
    cpuinfo = subprocess.check_output("lscpu", shell=True).strip().decode()
    
    chips = int(subprocess.check_output("lscpu | grep 'Socket(s):' | awk '{print $2}'", shell=True).strip())
    threads = int(subprocess.check_output("lscpu | grep 'Thread(s) per core:' | awk '{print $4}'", shell=True).strip()) * \
              int(subprocess.check_output("lscpu | grep 'Core(s) per socket:' | awk '{print $4}'", shell=True).strip()) * chips
    cores = int(subprocess.check_output("lscpu | grep 'Core(s) per socket:' | awk '{print $4}'", shell=True).strip()) * chips
    frequency = int(float(subprocess.check_output("lscpu | grep 'MHz' | awk '{print $3}'", shell=True).strip()))

    # Architecture
    architecture = subprocess.check_output("uname -m", shell=True).strip().decode()

    # CPU Make (This might not always give accurate results)
    cpu_make = subprocess.check_output("cat /proc/cpuinfo | grep 'model name' | uniq | awk -F ': ' '{print $2}'", shell=True).strip().decode()

    # Release Year and TDP are difficult to obtain programmatically
    release_year = "Unknown"
    tdp = "Unknown"

    # RAM
    meminfo = subprocess.check_output("cat /proc/meminfo | grep MemTotal", shell=True).strip().decode()
    ram = int(re.findall(r'\d+', meminfo)[0]) // 1024  # Convert to MB

    return {
        "CPU Chips": chips,
        "CPU Threads": threads,
        "CPU Cores": cores,
        "CPU Frequency (MHz)": frequency,
        "Architecture": architecture,
        "CPU Make": cpu_make,
        "Release Year": release_year,
        "RAM (MB)": ram,
        "TDP": tdp
    }

if __name__ == "__main__":
    info = get_cpu_info()
    for key, value in info.items():
        print(f"{key}: {value}")
```

Maybe we can use something like https://www.techpowerup.com/cpu-specs/ to find the TDP

### PR DESCRIPTION
**Pull request by ArneTR**
closes #21 

Implemented only CPU and RAM info for now.

We technically have all the TDP data in csv files and could scrape it from there. This involves bigger PR though

### CODE REVIEW

File: auto_detect.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        # we also currently do not matc for make, as this info can result in ARM which is currently not supported and
+        # would rather lead to confusion
+    #pylint: disable=broad-except
+    except Exception as err:
+        print_custom('Exception', err, silent=silent)
+        print_custom('Could not check for CPU info. Setting all values to None.', silent=silent)
+
+
+    try:
+        meminfo = subprocess.check_output(['cat', '/proc/meminfo'], encoding='UTF-8', stderr=subprocess.DEVNULL)
```
Comment (ribalba) [inline auto_detect.py:65]: Why do you pipe to devnull here?

**Review by ribalba — APPROVED** [review]
Looks good. Why not use std logging instead of the `print_custom` method? I would probably also warn if you are running this on a non linux machine. Just to make it clear that it is intended for Linux use cases.

**Comment by ArneTR** [comment]
> Looks good. Why not use std logging instead of the `print_custom` method? I would probably also warn if you are running this on a non linux machine. Just to make it clear that it is intended for Linux use cases.

that is not quite true. everyhere where `lscpu` is availabe this works. This can also be in a WSL2 or in a docker container on mac.

I think going the python approach and just reading the paths is reasonable. Making guesses *why* it does not work might be helpful, but can also be misleading (restricted VM for instance).

**Comment by ArneTR** [comment]
I sadly come to the understanding that the cpu freq detection is wrong.

The advertised frequency in lscpu is the max freq. The model uses the base freq though.

This is apparently only visible in the model name. 

A possible route could be to do it in the `/proc/cpuinfo` but I found the variable to be fluctuating greatly and not very helpful sometims. Still I implemented it as a fallback and called it "assumed".

Problem is that the assumption is that one core is always running at least at base frequency and thus using max() is reasonable. However I see machines even in idle quite often hanging in Turbo Boost for a little bit. This makes the value sometimes bogus.

**Comment by ArneTR** [comment]
@ribalba I added auto detection for make and TDP (through RAPL power capping)

Can you check on your linux system if you get correct values?

Do: `python3 auto_detect.py` and then afterwards just `python3 xgb.py`

In the lattter the auto_detect should be picked up

**Comment by ribalba** [comment]
So auto detect is picked up and the values are sort of ok. The problem is that I have performance and energy cores that are not detected properly. I don't think this is really a problem as we are mostly working on servers and spec power doesn't really account for these different core technologies. 

https://www.intel.de/content/www/de/de/products/sku/226253/intel-core-i71280p-processor-24m-cache-up-to-4-80-ghz/specifications.html

```
The following data was auto detected: {'freq': 4475, 'threads': 20, 'cores': 10, 'tdp': 28.0, 'mem': 32, 'make': 'intel', 'chips': 1}
```

**Comment by ArneTR** [comment]
> So auto detect is picked up and the values are sort of ok. The problem is that I have performance and energy cores that are not detected properly. I don't think this is really a problem as we are mostly working on servers and spec power doesn't really account for these different core technologies.
> 
> https://www.intel.de/content/www/de/de/products/sku/226253/intel-core-i71280p-processor-24m-cache-up-to-4-80-ghz/specifications.html
> 
> ```
> The following data was auto detected: {'freq': 4475, 'threads': 20, 'cores': 10, 'tdp': 28.0, 'mem': 32, 'make': 'intel', 'chips': 1}
> ```

If we wanted to drill deeper: In which tool in Linux do you get a better insight on the type of cores you have? Does lscpu or any other subsystem have more data that we can scrape?

**Comment by ribalba** [comment]
So I looked into 
```
$ dmidecode -t processor
...
	Core Count: 14
	Core Enabled: 14
	Thread Count: 20

```
and 
```
hwinfo
```
sort of gives me the information. But not very easy to parse as every core comes up as an own section.

```
$ cat /proc/cpuinfo
siblings	: 20
cpu cores	: 14
```
also gives me the correct info


but none of them really shows how

File: auto_detect.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+            data['chips'] = int(match.group(1))
+            logger.info('Found Sockets: %d (will take precedence if not 0)', data['chips'])
+        else:
+            logger.info('Could not find Chips/Sockets via lscpu')
+
+        if data['threads'] and data['chips']:
+            match = re.search(r'Thread\(s\) per core:\s*(\d+)', cpuinfo)
+            if match:
+                threads_per_core = int(match.group(1))
+                data['cores'] = (data['threads'] // threads_per_core) // data['chips']
```
Comment (ribalba) [inline auto_detect.py:88]: I think the problem is here that not every core has 2 threads on my computer. You could get the cores through getting 
```
    cores_per_socket_match = re.search(r'Core\(s\) per socket:\s*(\d+)', input_str)
    sockets_match = re.search(r'Socket\(s\):\s*(\d+)', input_str)


    if cores_per_socket_match and sockets_match:
        cores_per_socket = int(cores_per_socket_match.group(1))
        sockets = int(sockets_match.group(1))
        data['cores'] cores_per_socket * sockets
```

Comment (ArneTR) [reply auto_detect.py:88]: Can you please post the full dumps of `lscpu`, `hwinfo` and `/proc/cpuinfo`.

I cannot fully implement the regex without it.

**Comment by ribalba** [comment]
I need to do this at home. Sorry I didn't bring my Linux Laptop

File: auto_detect.py (hunk 2 above)
Comment (ribalba) [reply auto_detect.py:88]: [commands.txt](https://github.com/green-coding-berlin/spec-power-model/files/13808787/commands.txt)

Comment (ribalba) [reply auto_detect.py:88]: @ArneTR ping

**Comment by ArneTR** [comment]
ok, I have removed the derivation of the cores and will only scrape that value if it shows up in lscpu.

Also the frequency guessing according to current frequency is removed. Only using that if it shows up in the CPU name tag.

Can you give it another try? @ribalba

**Comment by ribalba** [comment]
So 
```
$ python3 auto_detect.py 
Found TDP: 28 W
Found Sockets: 1
Exception: Domain /sys/class/powercap/intel-rapl/intel-rapl:1/name was not package-1, but psys
Could not find (additional) chips info under file path. Most likely reached final chip. continuing ...
Found Threads: 20
Found Sockets: 1 (will take precedence if not 0)
Found cores: 14 
Could not find Frequency. Using default None
Found Make: intel
Found Memory: 32 GB
{'freq': None, 'threads': 20, 'cores': 14, 'tdp': 28.0, 'mem': 32, 'make': 'intel', 'chips': 1}
```
Everything looks ok except that freq now stopped working. I just pushed a fix for my machine.

**Comment by ArneTR** [comment]
No, that fix won't do. What you are getting is the max frequency. But we need the base frequency, as this is what comes with SPECPower.

I have reverted the patch and merged.

Sadly Intel does not even show the base frequency anymore on their data sheets: https://www.intel.de/content/www/de/de/products/sku/226253/intel-core-i71280p-processor-24m-cache-up-to-4-80-ghz/specifications.html

This was different in older generations: https://www.intel.de/content/www/de/de/products/sku/33910/intel-core2-duo-processor-e8400-6m-cache-3-00-ghz-1333-mhz-fsb/specifications.html?wapkw=intel%20core%202%20duo%20e8400

**Pull request merged**